fv/wireguard_test.go

```diff
@@ -0,0 +1,444 @@
+		}
+
+		// Create workloads, One on each "host".
+		for ii := range w {
+			wIP := fmt.Sprintf("10.65.%d.2", ii)
+			wName := fmt.Sprintf("w%d", ii)
+
+			err := client.IPAM().AssignIP(context.Background(), ipam.AssignIPArgs{
+				IP:       net.MustParseIP(wIP),
+				HandleID: &wName,
+				Attrs: map[string]string{
+					ipam.AttributeNode: felixes[ii].Hostname,
+				},
+				Hostname: felixes[ii].Hostname,
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			w[ii] = workload.Run(felixes[ii], wName, "default", wIP, "8055", "tcp")
+			w[ii].ConfigureInDatastore(infra)
+
+			routeEntryPattern[ii] = fmt.Sprintf("10.65.%d.0/26 dev %s scope link", ii, wireguardInterfaceNameDefault)
+		}
+		// create workload in "host" namespace.
+		for ii := range hostW {
+			hostW[ii] = workload.Run(felixes[ii], fmt.Sprintf("host%d", ii), "", felixes[ii].IP, "8055", "tcp")
+		}
+
+		cc = &connectivity.Checker{}
+
+		// to get route-table index from rule.
+		routingRuleRegExp = regexp.MustCompile(`\d+$`)
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			for _, felix := range felixes {
+				felix.Exec("ip", "link", "show")
+				felix.Exec("ip", "rule", "list")
+				felix.Exec("ip", "route", "show", "table", "all")
+				felix.Exec("ip", "route", "show", "cached")
+				felix.Exec("/wg")
+			}
+		}
+
+		for _, wl := range w {
+			wl.Stop()
+		}
+		for _, hwl := range hostW {
+			hwl.Stop()
+		}
+		for _, t := range tcpdumps {
+			t.Stop()
+		}
+		for _, felix := range felixes {
+			felix.Stop()
+		}
+
+		if CurrentGinkgoTestDescription().Failed {
+			infra.DumpErrorData()
+		}
+		infra.Stop()
+	})
+
+	tcpdumpMatchCount := func(felix int, matcherName string) func() int {
+		return func() int {
+			return tcpdumps[felix].MatchCount(matcherName)
+		}
+	}
+
+	Context("with Wireguard enabled", func() {
+		BeforeEach(func() {
+			for i, felix := range felixes {
+				// Check the Wireguard device exists.
+				Eventually(func() error {
+					out, err := felix.ExecOutput("ip", "link", "show", wireguardInterfaceNameDefault)
+					if err != nil {
+						return err
+					}
+					if strings.Contains(out, wireguardInterfaceNameDefault) {
+						return nil
+					}
+					return fmt.Errorf("felix %d has no Wireguard device", i)
+				}, "10s", "100ms").ShouldNot(HaveOccurred())
+			}
+		})
+
+		It("workload to workload traffic should be allowed and encrypted", func() {
+			cc.ExpectSome(w[0], w[1])
+			cc.ExpectSome(w[1], w[0])
+			cc.CheckConnectivity()
+
+			for i := range felixes {
+				Eventually(tcpdumpMatchCount(i, "numInWorkloadPackets")).Should(BeNumerically(">", 0))
+				Eventually(tcpdumpMatchCount(i, "numOutWorkloadPackets")).Should(BeNumerically(">", 0))
+				Eventually(tcpdumpMatchCount(i, "numTunnelOutboundPackets")).Should(BeNumerically(">", 0))
+				Eventually(tcpdumpMatchCount(i, "numTunnelInboundPackets")).Should(BeNumerically(">", 0))
+			}
+		})
+
+		It("the Wireguard rule-entry/route-table should exist", func() {
```

Comment:
Usually, I wouldn't check this level of detail (better to check the rules in UT and then just verify the behaviour here).  However, combined with the negated check when wg is disabled, maybe it's a good cross check.